bugFix - async bucket merge race

sisuresh opened the pull request:
# Description

Resolves #2393

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

marta-lokhova commented on src/catchup/CatchupWork.cpp:
```diff
@@ -343,9 +343,19 @@ CatchupWork::runCatchupStep()
             }
 
-            // Step 4.4: Apply buffered ledgers
+            auto onSuccessCallback = [&]() {
```
Because we're passing this callback to another class, safest is to only capture things you need, and avoid potential dangling references. Here you can capture ProgressState and catchup mode by value (they remain constant anyway), and have the callback take `ledgerSeq` argument instead of `mLastApplied` (you can deduce last applied ledger by asking LedgerManager). The only thing you need to capture by reference is the progress handler function.

sisuresh replied:
fixed

marta-lokhova commented on src/catchup/ApplyBufferedLedgersWork.cpp:
```diff
@@ -39,6 +41,7 @@ ApplyBufferedLedgersWork::onRun()
     auto& lm = mApp.getLedgerManager();
     if (!lm.hasBufferedLedger())
     {
+        mOnSuccess();
```
A cleaner way would be to do this callback in designated the `onSuccess` function. This way, if this function changes to return WORK_SUCCESS in multiple places, `onSuccess` would properly handle it.

sisuresh replied:
done

sisuresh commented on src/catchup/CatchupWork.h:
```diff
@@ -165,6 +164,7 @@ class CatchupWork : public Work
     WorkSeqPtr mBucketVerifyApplySeq;
     std::shared_ptr<Work> mTransactionsVerifyApplySeq;
     std::shared_ptr<BasicWork> mApplyBufferedLedgersWork;
+    std::shared_ptr<BasicWork> mTesrApplyBufferedLedgersWork;
```
removed

sisuresh commented on src/catchup/CatchupWork.cpp:
```diff
@@ -343,9 +343,19 @@ CatchupWork::runCatchupStep()
 
-            // Step 4.4: Apply buffered ledgers
+            auto onSuccessCallback = [&]() {
+                CLOG(INFO, "History") << "Catchup finished";
```
fixed

sisuresh commented:
@marta-lokhova I addressed the comments related to the ApplyBufferedLedgersWork success and LedgerManager state transition change. I'm looking into the second issue (potential out of order ledgers during catchup)

MonsieurNicolas commented:
It looks like there are still some races in this version: basically, it's possible to start a new Catchup work (because of buffered ledgers coming in) while the old Catchup Work is still running... which is problematic, worst, it's possible that the old Catchup work resets `mCatchup` in ledger manager, which will cause another Catchup Work to start (so 3 running at the same time, in various state).

I prototyped a different change: I suggest we do something like https://github.com/MonsieurNicolas/stellar-core/commit/599bfd2d7f9361dc7c30f195d870965d528797c1

MonsieurNicolas commented:
opened a PR with that other change https://github.com/stellar/stellar-core/pull/2402

marta-lokhova commented:
I don't think the scenario described is possible, but I agree that we should go with your approach -- it's much easier to reason about, so I think it's less likely that we get into weird edge cases with `SyncingLedgerChain`. Taking a look at #2402

MonsieurNicolas commented:
landing the other PR instead
